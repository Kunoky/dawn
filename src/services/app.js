import axios from '@/utils/request'

export const listPost = () => {
  return axios.get('/posts')
}

export const getUser = id => {
  return axios.get('/users/' + id)
}
