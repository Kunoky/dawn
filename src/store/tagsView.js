import { defineStore } from 'pinia'

export const useTagsViewStore = defineStore('tagsView', () => {
  const list = ref([{ name: 'Home', meta: { title: '首页', affix: true } }])
  const names = computed(() => list.value.map(i => i.name))

  const userStore = useUserStore()

  const add = route => {
    if (!route.name) return
    const routeInfo = userStore.keyMenu[route.name]
    if (!routeInfo) return
    if (list.value.some(i => i.name === route.name) || routeInfo.meta.noCatch) return
    list.value.push(routeInfo)
  }

  const route = useRoute()
  const router = useRouter()
  const close = idx => {
    const r = list.value.splice(idx, 1)[0]
    if (r.name === route.name) {
      router.push({ name: list.value[idx - 1].name })
    }
  }
  return {
    list,
    names,
    add,
    close,
  }
})
