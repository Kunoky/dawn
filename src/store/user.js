import { defineStore } from 'pinia'
// import * as service from '@/services/user'
import { setToken, removeToken } from '@/utils/auth'
import router, { dynamicRoutes } from '@/router'

export const useUserStore = defineStore('user', {
  state: () => {
    const userStr = localStorage.getItem('user')
    let user = null,
      permission = []
    if (userStr) {
      user = JSON.parse(userStr)
      permission = user.permission
    }
    return {
      user: user,
      permission: permission,
      loading: {
        login: false,
      },
    }
  },
  getters: {},
  actions: {
    async login(user) {
      try {
        this.loading.login = true
        if (user.username === 'admin' && user.password === 'admin') {
          const data = {
            name: 'Admin',
            username: 'admin',
            permission: [],
            token: 'token',
          }
          localStorage.setItem('user', JSON.stringify(data))
          this.user = data
          this.permission = data.permission
          setToken(data.token)
        } else {
          return { msg: '用户名或密码错误' }
        }
        // const data = await service.login(user)
        // this.user = data
        return true
      } catch (e) {
        console.error('login fail: ', e)
        return e
      } finally {
        this.loading.login = false
      }
    },
    logout(go2login) {
      removeToken()
      localStorage.removeItem('user')
      localStorage.removeItem('menuTree')
      const route = router.history.current
      // TODO 当前路由是否需要登录
      const authRequired = true
      if (go2login || authRequired) {
        router.push({
          path: '/login',
          query: { redirect: route.fullPath },
        })
      }
    },
    addDynamicRoutes() {
      // TODO 如果需要，将此代码移至权限获取处，并完善动态添加路由
      dynamicRoutes.forEach(i => {
        let hasAuth = true
        if (!hasAuth) return
        if (i.pName) {
          router.addRoute(i.pName, i)
        } else {
          router.addRoute(i)
        }
      })
    },
  },
})
