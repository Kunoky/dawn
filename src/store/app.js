import { defineStore } from 'pinia'
import * as service from '@/services/app'
import { loadLanguageAsync } from '@/i18nSetup'
import { arr2tree } from '@/utils/common'
import zhCn from 'element-plus/dist/locale/zh-cn'

export const useAppStore = defineStore('app', {
  state: () => {
    const menuStr = localStorage.getItem('menuTree')
    let menuTree = []
    if (menuStr) {
      menuTree = JSON.parse(menuStr)
    }
    return {
      posts: [],
      user: {},
      lang: 'zh-CN',
      menuTree,
      langs: [
        {
          value: 'zh-CN',
          label: '简体中文',
        },
        {
          value: 'en',
          label: 'English',
        },
      ],
      elLangs: {
        'zh-CN': {
          value: zhCn,
          name: 'zh-cn',
        },
        en: {
          value: null,
          name: 'en',
        },
      },
      loading: {
        posts: false,
        lang: false,
      },
    }
  },
  getters: {
    total: state => state.posts.length,
    getUserById: state => {
      return id => {
        if (!Object.hasOwn(state.user, id)) {
          state.getUser(id)
        }
        return state.user[id]
      }
    },
    elLang: state => state.elLangs[state.lang].value,
  },
  actions: {
    async listPost() {
      this.loading.posts = true
      try {
        const posts = await service.listPost()
        this.posts = posts
      } catch (e) {
        console.error('load post fail: ', e)
      } finally {
        this.loading.posts = false
      }
    },
    async getUser(id, force) {
      if (this.user[id] && !force) return
      this.user[id] = null
      try {
        const user = await service.getUser(id)
        this.user = { ...this.user, [id]: user }
      } catch (e) {
        console.error('load post fail: ', e)
      }
    },
    async setLang(lang) {
      this.loading.lang = true
      try {
        const loaders = [loadLanguageAsync(lang)]
        if (!this.elLangs[lang].value) {
          loaders.push(import(`../../node_modules/element-plus/dist/locale/${this.elLangs[lang].name}.mjs`))
        }
        const [, elLang] = await Promise.all(loaders)
        if (elLang) {
          this.elLangs[lang].value = elLang.default
        }
        this.lang = lang
      } finally {
        this.loading.lang = false
      }
    },
    genMenu() {
      const menu = [
        {
          id: 1,
          key: '/',
          val: '导航一',
          meta: '{"title":"导航一","icon":"message"}',
        },
        {
          id: 2,
          key: '/demo',
          val: 'demo',
          pId: 1,
        },
        {
          id: 3,
          key: '/path1',
          val: '二级-选项1',
          pId: 1,
        },
        {
          id: 4,
          key: '/path2',
          val: '二级-选项2',
          pId: 1,
        },
        {
          id: 5,
          key: '/path2/path3',
          val: '三级-选项3',
          pId: 4,
        },
        {
          id: 6,
          key: '/path2/path4',
          val: '三级-选项4',
          pId: 4,
        },
        {
          id: 7,
          key: 'https://cn.vitejs.dev/',
          val: 'Vite',
        },
        {
          id: 8,
          key: 'https://pinia.vuejs.org/zh/',
          val: 'Pinia',
        },
        {
          id: 9,
          key: 'https://vitest.dev/',
          val: 'Vitest',
        },
        {
          id: 10,
          key: 'https://github.com/vuejs/awesome-vue',
          val: 'awesome',
        },
      ]

      const [menuTree] = arr2tree(menu)
      this.menuTree = menuTree
      localStorage.setItem('menuTree', JSON.stringify(menuTree))
    },
  },
})
