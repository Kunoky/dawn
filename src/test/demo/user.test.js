import { mount } from '@vue/test-utils'
import TestComponent from '@/view/demo/components/User.vue'
import { setActivePinia, createPinia } from 'pinia'
import { useAppStore } from '@/store/app'

describe('测试demo', () => {
  let store
  beforeAll(() => {
    /**单元测试环境下，需要显式的激活pinia */
    setActivePinia(createPinia())
    store = useAppStore()
  })

  it('mount component', async () => {
    expect(TestComponent).toBeTruthy()
    const id = 1
    await store.getUser(id)
    const wrapper = mount(TestComponent, {
      propsData: {
        id,
      },
    })
    expect(wrapper.text()).toContain('Leanne Graham')
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.get('b').trigger('click')

    expect(wrapper.text()).toContain('1-770-736-8031 x56442')
  })
})
