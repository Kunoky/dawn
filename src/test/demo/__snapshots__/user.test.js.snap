// Vitest Snapshot v1

exports[`测试demo > mount component 1`] = `"<span><b>Leanne Graham</b><span>1-770-736-8031 x56442</span></span>"`;
